### ISSUE TITLE
SSL certificates incorrectly extracted

### ISSUE THREAD

**Issue by explosivose** [ntop/nDPI#97]
### Context:

I'm using some TLSv1 test data with no client authenticated handshake so to my understanding there should be no SSL client certificate, and an SSL server certificate. 
### My observation:

`flow->protos.ssl.client_certificate` is populated using a Client Hello message in TLSv1 which, to my understanding, is incorrect, as it if the server that is providing the certificate later in the handshake. 

https://en.wikipedia.org/wiki/Transport_Layer_Security#Basic_TLS_handshake

The result is that `flow->protos.ssl.client_certifcate` is incorrectly populated with the server hostname from the Hello message and the server certificate names are missed from the Server Certificate message. 

https://github.com/ntop/nDPI/blob/dev/src/lib/protocols/ssl.c#L216

**Comment by kYroL01** [comment]
Hello @explosivose .
You say in the dissector is missing the separation between the case where a client and a server send both the certificate, and the case where only the server must send, if I correctly understand.

**Comment by explosivose** [comment]
My understanding of SSL is very limited

I don't expect any certificates to be found in the Client Hello message - is this a correct assumption?

**Comment by kYroL01** [comment]
Sorry for anwer in late.

Well, the ssl dissector is correct.
I explain better.
When nDPI encounter for the first time the _Client Hello_ packet, the function involved is not the one you highlited, but the **ndpi_search_ssl_tcp** (https://github.com/ntop/nDPI/blob/dev/src/lib/protocols/ssl.c#L536).
This function, first of all, try to understand if the packet analized is an SSL or Whattsapp pkt; if it's an SSL, call the function **sslDetectProtocolFromCertificate** that (as the name help) try to understand the protocol from the certificate. If the packet is a Client Hello, the function return immediately, and we try to understand if the packet is a SSLv[n] from server or client.
After this, the dissector return to the main function and the following packet is examined.
Only in case we have the complete certificate (after the exchange between Client and Server) the function you called is populated, but in that case we are sure the Client has exchanged is own certificate with the server.

In conclusion, that function is not called always during a SSL/TLS dissection. If yes, is called only in case we are sure that also the Client has the certificate.

If your doubts are clarified, I'll closed this issue.
If you have other questions, please ask.
If you have pcap that not confirm this environment of ssl dissection, please, provide us a copy.
Thank you.
don't forget to star nDPI ;)

**Comment by lmangani** [comment]
Closing for inactivity, reopen if needed.
